htdocs/contact/card.php

```diff
@@ -441,7 +429,7 @@
 
 					$object->old_lastname = '';
 					$object->old_firstname = '';
-					$action = 'view';
+					$action = 'update_extras';
```

Comment:
I think here you set update_extras to make the update of product in 2 steps. Firs you update the product and then later, in the if $action == "update_extras" you will update extra fields (so you reuse the controller code added to manage the update of extrafields alone). But there is a problem. If backtopage is defined, there is a redirect before we reach the part that update the update_extras. Also this means we update the extrafields into another transaction than the main properties. Also it means that the trigger inside the update of product used by external module will be run on a product that is not yet completely updated.
For this reason, you must keep the code juste after the comment
// Fill array 'array_options' with data from add form
So extrafields is filled and the call of  $object->update  will both update comm properties and extrafields (in same transaction, including trigger)

I did not test but i think your add 
if ($action == "update_extras" && !empty($permissiontoadd)) {
should be enough to allow edition of extrafields when we update them alone.